File: backend/src/main/java/com/coing/infra/upbit/util/UpbitRequestBuilder.java
package com.coing.infra.upbit.util;

import java.util.Arrays;
import java.util.List;

import org.springframework.stereotype.Component;

import com.coing.domain.coin.market.entity.Market;
import com.coing.domain.coin.market.service.MarketCacheService;
import com.coing.infra.upbit.dto.UpbitWebSocketFormatDto;
import com.coing.infra.upbit.dto.UpbitWebSocketTicketDto;
import com.coing.infra.upbit.dto.UpbitWebSocketTypeDto;
import com.coing.infra.upbit.enums.EnumUpbitRequestType;
import com.coing.infra.upbit.enums.EnumUpbitWebSocketFormat;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class UpbitRequestBuilder {
	private final ObjectMapper objectMapper = new ObjectMapper();
	private final MarketCacheService marketCacheService;

	public String makeRequest(EnumUpbitRequestType type) throws JsonProcessingException {
		List<String> codes = marketCacheService.getCachedMarketList().stream().map(Market::getCode).toList();
		UpbitWebSocketTicketDto ticketDto = UpbitWebSocketTicketDto.builder()
			.ticket(type.getValue())
			.build();
		UpbitWebSocketTypeDto typeDto = UpbitWebSocketTypeDto.builder()
			.type(type.getValue())
			.codes(codes)
			.isOnlyRealtime(false)
			.isOnlySnapshot(false)
			.build();
		UpbitWebSocketFormatDto formatDto = UpbitWebSocketFormatDto.builder()
			.format(EnumUpbitWebSocketFormat.SIMPLE)
			.build();

		List<Object> dataList = Arrays.asList(ticketDto, typeDto, formatDto);
		return objectMapper.writeValueAsString(dataList);
	}
}

File: backend/src/main/java/com/coing/infra/upbit/util/UpbitRequestBuilder.kt
package com.coing.infra.upbit.util

import com.coing.domain.coin.market.service.MarketCacheService
import com.coing.infra.upbit.adapter.websocket.dto.UpbitWebSocketFormatDto
import com.coing.infra.upbit.adapter.websocket.dto.UpbitWebSocketTicketDto
import com.coing.infra.upbit.adapter.websocket.dto.UpbitWebSocketTypeDto
import com.coing.infra.upbit.adapter.websocket.enums.EnumUpbitWebSocketFormat
import com.coing.infra.upbit.adapter.websocket.enums.EnumUpbitWebSocketRequestType
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.stereotype.Component
import java.util.*

@Component
class UpbitRequestBuilder(
    private val marketCacheService: MarketCacheService,
    private val objectMapper: ObjectMapper = ObjectMapper()
) {

    @Throws(JsonProcessingException::class)
    fun makeWebSocketRequest(type: EnumUpbitWebSocketRequestType): String {
        val codes = marketCacheService.getCachedMarketMap().keys.toList()
        val ticketDto = UpbitWebSocketTicketDto(
            ticket = type.value
        )
        val typeDto = UpbitWebSocketTypeDto(
            type = type.value,
            codes = codes,
            isOnlyRealtime = false,
            isOnlySnapshot = false
        )
        val formatDto = UpbitWebSocketFormatDto(
            format = EnumUpbitWebSocketFormat.SIMPLE
        )

        val dataList = listOf(ticketDto, typeDto, formatDto)
        return objectMapper.writeValueAsString(dataList)
    }
}
